### PR TITLE
Named templates

### PR DESCRIPTION
**Pull request by Aden-git**
Improve named templates support.
Part of the issue #79 fix.
- Implement the DevExtreme DevExpress.ui.TemplateBase interface
- Split the spec file into two parts

### CODE REVIEW

File: tests/src/template.spec.ts [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+
+                expect(templatesHash['testTemplate']).not.toBeUndefined();
+                expect(typeof templatesHash['testTemplate'].render).toBe('function');
+
+                done();
+            })
+            .catch(e => done.fail(e));
+    });
+
+    it('should have methods render, dispose, owner and source in template', done => {
```
Comment (dxbykov) [inline tests/src/template.spec.ts]: I would name this test as 'should implement the DevExpress.ui.TemplateBase interface'

File: tests/src/template.spec.ts [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+                expect(typeof template.dispose).toBe('function');
+                expect(typeof template.owner).toBe('function');
+                expect(typeof template.source).toBe('function');
+
+                done();
+            })
+            .catch(e => done.fail(e));
+    });
+
+    it('should have implementation of methods render, dispose, owner and source in template', done => {
```
Comment (dxbykov) [inline tests/src/template.spec.ts]: By name looks like a duplicate of the previous test :)

File: tests/src/template.spec.ts [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+                fixture.detectChanges();
+
+                let instance = getWidget(fixture),
+                    templatesHash = instance.option('_templates'),
+                    template = templatesHash['testTemplate'],
+                    itemData = {},
+                    itemIndex = 0,
+                    itemElement = $('<div>');
+
+                expect(typeof template.render(itemData, itemIndex, itemElement)).toBe('object');
```
Comment (dxbykov) [inline tests/src/template.spec.ts]: Isn't this assert just one more case from the test below ('should nonrmalize template function arguments order (#17)')?

File: tests/src/template.spec.ts [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+                expect(typeof template.owner()).toBe('object');
+                expect(typeof template.source()).toBe('object');
+                template.dispose();
```
Comment (dxbykov) [inline tests/src/template.spec.ts]: If we need to check if the implementation works properly then type checking is not enough. Let's check the methods return objects we expect.

File: tests/src/template.spec.ts (hunk 3 above)
Comment (Aden-git) [inline tests/src/template.spec.ts]: No, this assert just check a render function.locate in test, named "should nonrmalize template function arguments order (#17)"

Comment (Aden-git) [inline tests/src/template.spec.ts]: No, this assert check a result of "render" function.
As for as new case for test 'should nonrmalize template function arguments order (#17)', so it was added there.

File: tests/src/template.spec.ts (hunk 2 above)
Comment (Aden-git) [inline tests/src/template.spec.ts]: Seems it to be true. I'll remove it)

File: tests/src/template.spec.ts (hunk 1 above)
Comment (Aden-git) [inline tests/src/template.spec.ts]: You a right about name, now it seems obvious.

To create a DxTemplate instance it needs an instances of TemplateRef, ViewContainerRef and DxTemplateHost. So I think, thet more easy get access to the DxTemplate instance from widget.

File: tests/src/template.spec.ts (hunk 4 above)
Comment (Aden-git) [inline tests/src/template.spec.ts]: I'm agree with you, it will be better. I'll fix it.

**Pull request merged**